### PR TITLE
State achievements that depend on assessments in assessment listings

### LINKED ISSUE #1044: Add achievements that depend on each assessment
For each assessment at the course homepage. show achievements that will be unlocked when the assessment is completed.

### PR DESCRIPTION
**Pull request by weiqingtoh**
Fixes #1044 

To prevent N+1, I used a service to preload all the assessments and achievements (along with the conditions and conditionals).

### CODE REVIEW

File: app/services/course/assessment/achievement_condition_service.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  # Loads the relevant achievements
+  def achievements
+    @achievements ||= begin
+      achievement_ids = assessment_achievement_hash.values.flatten.uniq
+      Course::Achievement.where(id: achievement_ids)
+    end
+  end
+
+  # Builds the hash linking the specific assessment_id to the achievement_id.
+  # eg. [[1,2], [1,4], [3,4]] Indicates assessment 1 is required for achievements 2 and 4, while
```
Comment (fonglh) [inline app/services/course/assessment/achievement_condition_service.rb]: If it's a hash, should be represented as `{1: [2, 4], 3: [4]}` instead

Comment (weiqingtoh) [inline app/services/course/assessment/achievement_condition_service.rb]: Thanks 👍 missed this out!

File: app/services/course/assessment/achievement_conditional_service.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+# This service preloads all Achievement conditionals which lists Assessments as conditions.
+# Used for Assessments#Index to reduce n+1 queries.
+class Course::Assessment::AchievementConditionalService
+  # Initialises the service with the listed assessments.
+  #
+  # @param [Array<CourseUser>] course_users
```
Comment (allenwq) [inline app/services/course/assessment/achievement_conditional_service.rb]: Should be assessments ?

File: app/services/course/assessment/achievement_conditional_service.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+# This service preloads all Achievement conditionals which lists Assessments as conditions.
+# Used for Assessments#Index to reduce n+1 queries.
+class Course::Assessment::AchievementConditionalService
```
Comment (allenwq) [inline app/services/course/assessment/achievement_conditional_service.rb]: How about call this `AchievementPreload` service ? I think current name is too general and does not match is purpose.

I remember we have a service to preload tutors, maybe it's good to call it `TutorPreload` service too, @kxmbrian What you think ?

Comment (kxmbrian) [inline app/services/course/assessment/achievement_conditional_service.rb]: That would work 👍

**Pull request merged**